### PR TITLE
[master] Porting #50696 to master

### PR DESCRIPTION
**Pull request by garethgreenaway**
Porting #50696 to master

### CODE REVIEW

**Comment by garethgreenaway** [comment]
re-run full all

File: salt/modules/saltutil.py [hunk 1]
```diff
@@ -1444,6 +1445,14 @@ def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs):
...
         }
         del kwargs['subset']
+    elif kwargs.get('asynchronous'):
+        cmd_kwargs = {
+            'tgt': tgt, 'fun': fun, 'arg': arg, 'tgt_type': tgt_type,
+            'ret': ret, 'kwarg': kwarg
+        }
+        del kwargs['asynchronous']
+        # run_job doesnt need processing like the others
+        return client.run_job(**cmd_kwargs)
```
Comment (dhiltonp) [inline salt/modules/saltutil.py]: Passed kwargs aren't ever applied to cmd_kwargs, unlike the other cases. Keeping consistent behavior would be good.

A thing I'm not clear on - why does this chunk bypass calling `_cmd` like the other ways the function is called?

File: salt/modules/saltutil.py [hunk 2]
```diff
@@ -1444,6 +1445,14 @@ def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs):
             'ret': ret, 'cli': True, 'kwarg': kwarg, 'sub': kwargs['subset'],
         }
         del kwargs['subset']
+    elif kwargs.get('asynchronous'):
```
Comment (dhiltonp) [inline salt/modules/saltutil.py]: We should change the function signature to:

`def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, batch=False, subset=False, asynchronous=False, **kwargs):`

That way we don't need to remove the flags from kwargs manually - any excess options will be passed to the command.

File: salt/modules/saltutil.py [hunk 3]
```diff
@@ -1444,6 +1445,14 @@ def _exec(client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs):
...
+        cmd_kwargs = {
```
Comment (dhiltonp) [inline salt/modules/saltutil.py]: Maybe we should define `cmd_kwargs` above the if statements, then just modify it with the differences in the respective if clauses.

On a related note, are the clauses `batch`, `subset` and `asynchronous` mutually exclusive? @waynew what do you think?

File: salt/modules/saltutil.py [hunk 4]
```diff
@@ -1608,11 +1622,16 @@ def runner(name, arg=None, kwarg=None, full_return=False, saltenv='base', jid=No
...
 
-    return rclient.cmd(name,
-                       arg=arg,
-                       kwarg=kwarg,
-                       print_event=False,
-                       full_return=full_return)
+    if asynchronous:
+        master_key = salt.utils.master.get_master_key('root', __opts__)
+        low = {'arg': arg, 'kwarg': kwarg, 'fun': name, 'key': master_key}
+        return rclient.cmd_async(low)
```
Comment (dhiltonp) [inline salt/modules/saltutil.py]: `rclient.cmd` and `rclient.cmd_async` seem to have inconsistent interfaces - a quick check and `LocalClient.cmd_async` accepts full parameters, but other `cmd_async`s accept a dictionary...

File: salt/modules/saltutil.py (hunk 2 above)
Comment (dhiltonp) [reply salt/modules/saltutil.py]: Doing that would also require removing a few lines from `saltmod.py:function` which determines if batch, subset and asynchronous are in kwargs.

**Comment by DmitryKuzmenko** [comment]
@garethgreenaway do you plan to follow @dhiltonp recommendations in scope of this PR?

**Comment by dhiltonp** [comment]
I've cleaned up the `_exec` implementation in #56203, finding a bug along the way.

It doesn't include this PR, though.

File: salt/modules/saltutil.py (hunk 4 above)
Comment (dhiltonp) [reply salt/modules/saltutil.py]: What's the deal with `low`? Is it normal elsewhere or something?

**Comment by dhiltonp** [comment]
I've merged the fixes I suggested with this code into master: #56480

**Comment by DmitryKuzmenko** [comment]
Thank you @dhiltonp

**Comment by dhiltonp** [comment]
I just realized I miscommunicated. That PR hasn't yet been merged, but it is based on current master and combines the two PRs.

**Comment by DmitryKuzmenko** [comment]
@dhiltonp this is how I got you. I've checked your PR and found it containing everything from this one. Am I wrong in closing this?

**Pull request closed without merging**